service/domain/downloader/scheduler.go

```diff
@@ -155,6 +157,16 @@ func newTaskSubscription(ctx context.Context, ch chan Task) *taskSubscription {
 	}
 }
 
+// A RelayTaskGenerator is responsible for generating tasks for a single relay.
+// Each task provides the time window for each query (since and until) and keeps
+// track of how many available running queries we can perform on the relay, we only
+// allow 3 running queries at a time; for global, author and tag queries across all
+// subscription channels.
+
+// RelayTaskGenerator maintains 3 TimeWindowTaskGenerator, one for each query
+// type. Each TimeWindowTaskGenerator maintains a list of TimeWindowTaskTracker,
+// one for each time window. Each TimeWindowTaskTracker maintains a list of
+// runningRelayDownloader, one for each concurrency setting. Each TimeWindowTaskTracker uses a TimeWindow
```

Comment:
Interesting. We have very similar code to this in the iOS app.

Reply:
Maybe this should be extracted into a library at some point